Add localResponseNormalization

chaosmail opened the pull request:
Hi guys,

I added CPU and GPU implementation of LRN (localResponseNormalization) for Array3D. I added docs and tests, more docs can be found here: http://caffe.berkeleyvision.org/tutorial/layers/lrn.html.

Best,
Christoph

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/517)
<!-- Reviewable:end -->

nsthorat commented:
Really nice work! A few comments inline.

Two high level comments:
* Could you do a really quick test with tf.local_response_normalization for some random input to make sure it matches TF since this is a non-trivial op? This is how we've been sanity checking our correctness.
* Hopefully the clang-format update will help you. Did you ever figure out what was going on there?

---

Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/math/lrn_test.ts, line 25 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-CUusiTw4zktO2Yt:-L2C-CUusiTw4zktO2Yu:b-2zp2sq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L25)):*
> ```XML
> // math.localResponseNormalization3D
> {
>   const sqArr = (arr: number[]) => arr.map((d) => Math.pow(d, 2));
> ```

you can remove () around d

`const sqArr = (arr: number[]) => arr.map(d => Math.pow(d, 2));`

---

*[src/math/lrn_test.ts, line 37 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-KAs6RyTXMyTIHh3:-L2C-KAs6RyTXMyTIHh4:bkrws9y) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L37)):*
> ```XML
>       const k = 1;
>       
>       const result = math.localResponseNormalization3D(x, n, alpha, beta, null,
> ```

can you pull null out to a variable and name it (so it's clear what you are nulling)

---

*[src/math/lrn_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-gM32eFHaeFZqc0a:-L2C-gM45vitC9qL18Zd:b-xk9cii) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L238)):*
> ```XML
>               x.get(1, 2, d1), x.get(2, 2, d1)),
>           ]);
>     });
> ```

can you add a couple unit tests for your throw conditions?

---

*[src/math/math.ts, line 2251 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2ByWtyZ_eNHndMu65G:-L2ByWtzBT1lTxEwfoZ9:b-sb0nls) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2251)):*
> ```XML
>    * This is used for Caffe compatibility only. New models should use BatchNorm
>    * instead!
>    * @param x he input NDArray.
> ```

s/he/the

---

*[src/math/math.ts, line 2259 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BxuCLstfrV-BqRLJW:-L2BxuCLstfrV-BqRLJX:b-fcx5ul) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2259)):*
> ```XML
>    * @param k A constant added to the basis.
>    */
>   localResponseNormalization3D(
> ```

can you make sure the arguments match TF, with the same names and the same order?

https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization

n => depth_radius
k => bias
beta default => 0.5

It looks like TensorFlow also accepts a batched Array4D, can we support that while we are here?

---

*[src/math/backends/backend.ts, line 165 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2By_EaQZWrYlSc7sMU:-L2By_EaQZWrYlSc7sMV:b6zgi48) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/backend.ts#L165)):*
> ```XML
>   localResponseNormalization3D(
>     x: Array3D, n: number, alpha: number, beta: number,
>     normRegion: "acrossChannels"|"withinChannel", k: number): Array3D;
> ```

Did you ever get the clang-format working? Auto save should change double quotes => single quotes

---

*[src/math/backends/backend_cpu.ts, line 1406 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BydL0EPrkDuWXaDQ2:-L2BydL0EPrkDuWXaDQ3:b-jd9xbc) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/backend_cpu.ts#L1406)):*
> ```XML
>           const sum = normRegion === "withinChannel"
>             ? sumWithinChannel(r, c, d)
>             : sumAcrossChannels(r, c, d);         
> ```

trailing space (hopefully clang-format fixes these errors)

---

*[src/math/backends/kernel_registry.ts, line 307 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Bz1AzWk_CEzFgBpnh:-L2Bz1AzWk_CEzFgBpni:bet0mxg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/kernel_registry.ts#L307)):*
> ```XML
>         config.args.varianceEpsilon, config.inputs.scale, config.inputs.offset);
>   },
>   LRN3D: (backend: MathBackend, config: LRN3DInputConfig) => {
> ```

if we make this kernel 4D while we are here, you wont need a 3D implementation (it will just call 4D with an extra 1 in the first dimension).

We did something very similar for batchnorm (the shader actually didnt change when we went 3D > 4D).

---

*[src/math/backends/webgl/lrn_gpu.ts, line 28 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzE_dyZxgXIN5f7wg:-L2BzE_dyZxgXIN5f7wh:b-527fzi) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L28)):*
> ```XML
>       xShape: number[], n: number, alpha: number, beta: number,
>       normRegion: "acrossChannels"|"withinChannel", k: number) {
>     
> ```

remove this empty line

---

*[src/math/backends/webgl/lrn_gpu.ts, line 57 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzzLcngUlfmGgFAsB:-L2BzzLcngUlfmGgFAsC:b-wju85n) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L57)):*
> ```XML
>               int idx = r + u;
>               int idy = c + v;
>               if (idx >= 0 && idx <= ${maxW} && idy >= 0 && idy <= ${maxH}) {
> ```

can you move bounds to the bounds of the for loop?

---

*[src/math/backends/webgl/lrn_gpu.ts, line 68 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Bzwndl0Xe-JGiHn9x:-L2Bzwndl0Xe-JGiHn9y:b-hs8rl4) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L68)):*
> ```XML
>       `;
>     }
>     else {
> ```

else should be on the previous line (clang-format will fix this)

---

*[src/math/backends/webgl/lrn_gpu.ts, line 84 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzmApNf_0b9M-iFyK:-L2BzmApNf_0b9M-iFyL:b-v3eu03) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L84)):*
> ```XML
>           for (int j = -${n2}; j <= ${n2}; j++) {
>             int idx = d + j;
>             if (idx >= 0 && idx <= ${maxD}) {
> ```

can you move this bounds check to the for loop bounds like you do in CPU? it will make this for loop much faster (actualy glsl will run the contents of the if statement always and then throw it out).

---

*[src/math/backends/webgl/lrn_gpu.ts, line 95 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzvrR_SN8ZMbHHLC0:-L2BzvrR_SN8ZMbHHLC1:bm8fxeh) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L95)):*
> ```XML
>     }
> 
>     return;
> ```

no need for return here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2C-w0v7EwR-G6_1kQg:b-qy4bl9)*
<!-- Sent from Reviewable.io -->

chaosmail commented:
Good hint to  compare with tf.local_response_normalization, it seems there are some subtle differences with parameters (e.g. bias vs k / n, beta vs -beta) etc. I have added to tests comparing with random tensors from TF.

---

Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/math/lrn_test.ts, line 25 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-CUusiTw4zktO2Yt:-L2CNzleGm_cnhm04lHJ:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L25)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

you can remove () around d

`const sqArr = (arr: number[]) => arr.map(d => Math.pow(d, 2));`
</blockquote></details>

Done.

---

*[src/math/lrn_test.ts, line 37 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-KAs6RyTXMyTIHh3:-L2CO24vka8Um9bQoZrN:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L37)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you pull null out to a variable and name it (so it's clear what you are nulling)
</blockquote></details>

Done.

---

*[src/math/lrn_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-gM32eFHaeFZqc0a:-L2CSltVVE9yQN9aiIdb:bc3tvdy) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L238)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you add a couple unit tests for your throw conditions?
</blockquote></details>

I added one for checking that the radius is odd. Is there any other invalid case or edge case I should test?

---

*[src/math/math.ts, line 2251 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2ByWtyZ_eNHndMu65G:-L2COkVuD14etqJn2YhV:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2251)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/he/the
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2259 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BxuCLstfrV-BqRLJW:-L2COCXQIaB9YnRdf5HU:b-rcyfzc) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2259)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you make sure the arguments match TF, with the same names and the same order?

https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization

n => depth_radius
k => bias
beta default => 0.5

It looks like TensorFlow also accepts a batched Array4D, can we support that while we are here?
</blockquote></details>

renamed k => bias and beta default => 0.5

TF supports only "acrossChannels" mode and hence depth_radius makes sense here. However, we also support "withinChannel" and hence this radius applies also to spatial dimension.
Hence, I renamed n => radius.

---

*[src/math/backends/backend.ts, line 165 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2By_EaQZWrYlSc7sMU:-L2COpdSxL6NyPqR4Tpp:btfx9b2) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/backend.ts#L165)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Did you ever get the clang-format working? Auto save should change double quotes => single quotes
</blockquote></details>

clang-format-6.0 (current dev version) works ;)

---

*[src/math/backends/backend_cpu.ts, line 1406 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BydL0EPrkDuWXaDQ2:-L2CR7papPDkkRN2Q002:b-8yam15) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/backend_cpu.ts#L1406)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

trailing space (hopefully clang-format fixes these errors)
</blockquote></details>

now it looks better with clang-format active.

---

*[src/math/backends/kernel_registry.ts, line 307 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Bz1AzWk_CEzFgBpnh:-L2CRC63Lb4ynebltcvj:b-hijxig) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/kernel_registry.ts#L307)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

if we make this kernel 4D while we are here, you wont need a 3D implementation (it will just call 4D with an extra 1 in the first dimension).

We did something very similar for batchnorm (the shader actually didnt change when we went 3D > 4D).
</blockquote></details>

Done. I had to tweak it a bit, as i need to access custom depth and spatial dims via `getX()`, but I think it is as you wanted it.

---

*[src/math/backends/webgl/lrn_gpu.ts, line 28 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzE_dyZxgXIN5f7wg:-L2CRROlQAiQRCAI9V9d:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L28)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove this empty line
</blockquote></details>

Done.

---

*[src/math/backends/webgl/lrn_gpu.ts, line 57 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzzLcngUlfmGgFAsB:-L2CSN8gncD0F3qloUEG:bk737q3) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L57)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you move bounds to the bounds of the for loop?
</blockquote></details>

I would love to, but it doesn't work. See non constant loop params restriction in the comment above.

---

*[src/math/backends/webgl/lrn_gpu.ts, line 68 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Bzwndl0Xe-JGiHn9x:-L2CS6qy4OJybyN2c77G:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L68)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

else should be on the previous line (clang-format will fix this)
</blockquote></details>

Done.

---

*[src/math/backends/webgl/lrn_gpu.ts, line 84 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzmApNf_0b9M-iFyK:-L2CRYxwV5Th5h4pd4QI:b-kuw9wr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L84)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you move this bounds check to the for loop bounds like you do in CPU? it will make this for loop much faster (actualy glsl will run the contents of the if statement always and then throw it out).
</blockquote></details>

Sure, that was my intention. However, OpenGL ES 2 doen't allow non constant loop params (Source: https://www.khronos.org/webgl/public-mailing-list/public_webgl/1012/msg00063.php)

---

*[src/math/backends/webgl/lrn_gpu.ts, line 95 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzvrR_SN8ZMbHHLC0:-L2CSFpcipgbDpgXUdSJ:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L95)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

no need for return here
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2CptvaOKb2TYEglvUT:b-29jabu)*
<!-- Sent from Reviewable.io -->

chaosmail commented:
Now we have localResponseNormalization for 3D and 4D including GPU optimization for `pow()` (source: [tensorflow/core/kernels/mkl_lrn_op.cc#L320][tf]) and tests to assure the same output for tf and dljs.

[tf]: https://github.com/tensorflow/tensorflow/blob/26033a1644a9c4a5fbe3170ab2e864b6a4ccd4ca/tensorflow/core/kernels/mkl_lrn_op.cc#L320

nsthorat commented:
Wonderful change! Mostly nits but one high level comment is to just have a 4D kernel, and have the 3D kernel expand to  a batch size of 4 and call the 4D op (not the kernel directly). Once you do this, this PR will simplify a ton!

Thanks for being thorough with this :)

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/math/lrn_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-gM32eFHaeFZqc0a:-L2GtqROoOR_iMDOSsEe:b-gw2lsx) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L238)):*
<details><summary><i>Previously, chaosmail wrote…</i></summary><blockquote>

I added one for checking that the radius is odd. Is there any other invalid case or edge case I should test?
</blockquote></details>

Yeah, asserting that the wrong rank fails (you can cast an Array2D as any so it compiles but gives you a runtime error)

---

*[src/math/math.ts, line 2259 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BxuCLstfrV-BqRLJW:-L2Gvn53t9sq2t6OLwFy:b-hebyu6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2259)):*
<details><summary><i>Previously, chaosmail wrote…</i></summary><blockquote>

renamed k => bias and beta default => 0.5

TF supports only "acrossChannels" mode and hence depth_radius makes sense here. However, we also support "withinChannel" and hence this radius applies also to spatial dimension.
Hence, I renamed n => radius.
</blockquote></details>

Can you make a note of this in the tsdoc?

Thanks :)

---

*[src/math/math.ts, line 2422 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Gu6ZpVMClBsL7RYxL:-L2Gu6ZpVMClBsL7RYxM:bx1hm2b) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/math.ts#L2422)):*
> ```XML
>          but got radius ${radius}.`);
> 
>     return this.backendEngine.executeKernel(
> ```

Let's kill the LRN3D kernel entirely. From the LRN3D op, we can reshape, then call math.LRN4D (don't call executeKernel on the same kernel from 2 places)

---

*[src/math/backends/backend_cpu.ts, line 1414 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2GuG1TCpVVamD3_4ze:-L2GuG1TCpVVamD3_4zf:bflj6r) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/backends/backend_cpu.ts#L1414)):*
> ```XML
>   }
> 
>   localResponseNormalization4D(
> ```

Once you only have a 4D kernel you can remove this shim, the 3D will call the 4D op

---

*[src/math/backends/types/lrn.ts, line 25 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Gvea5vYdCywihAgXF:-L2Gvea6vMdsI9Yj-4FB:b-jggaai) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/backends/types/lrn.ts#L25)):*
> ```XML
> 
> // 3D
> export interface LRN3DNode extends KernelNode {
> ```

you will only need the 4D node once you remove the 3D kernel

---

*[src/math/backends/webgl/lrn_gpu.ts, line 84 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzmApNf_0b9M-iFyK:-L2GuiLR1ah-sZ0bExj-:b-84x1nx) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L84)):*
<details><summary><i>Previously, chaosmail wrote…</i></summary><blockquote>

Sure, that was my intention. However, OpenGL ES 2 doen't allow non constant loop params (Source: https://www.khronos.org/webgl/public-mailing-list/public_webgl/1012/msg00063.php)
</blockquote></details>

Shouldn't the bounds of this be a function of arguments you splice in and not a function of anything at runtime?

---

*[src/math/backends/webgl/lrn_gpu.ts, line 20 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2GudpROKCRoooXOqHD:-L2GudpROKCRoooXOqHE:bf3fk5k) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/backends/webgl/lrn_gpu.ts#L20)):*
> ```XML
> import {GPGPUProgram} from './gpgpu_math';
> 
> export class LRNProgram implements GPGPUProgram {
> ```

this shader will get much simpler when we only support 4D input! :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2GvrbjBqz3u3QFnVq9:b-utpwnj)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
*batch size of 1

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2Gw6SYls5tYEdWmL-t:b-n38d0k)*
<!-- Sent from Reviewable.io -->

chaosmail commented:
This expanding and running on 4D is fantastic, makes the code very clean.

Is there a better way to expand and reduce the batch dimension than this:
```js
return this.localResponseNormalization4D(
                   x.reshape([1].concat(x.shape)) as Array4D, radius, bias,
                   alpha, beta, normRegion)
               .reshape(x.shape) as Array3D;
```

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[src/math/lrn_test.ts, line 238 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2C-gM32eFHaeFZqc0a:-L2HLS_EQy-MmVArTtcc:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/lrn_test.ts#L238)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Yeah, asserting that the wrong rank fails (you can cast an Array2D as any so it compiles but gives you a runtime error)
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2259 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BxuCLstfrV-BqRLJW:-L2HZED2aMLAlbEzoEUF:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/math.ts#L2259)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Can you make a note of this in the tsdoc?

Thanks :)
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2422 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2Gu6ZpVMClBsL7RYxL:-L2HZFm7B3Q7-QxT6jCG:bpjudz5) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/math.ts#L2422)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Let's kill the LRN3D kernel entirely. From the LRN3D op, we can reshape, then call math.LRN4D (don't call executeKernel on the same kernel from 2 places)
</blockquote></details>

Man, now I got you. This change is a beauty, thanks for the hint!

---

*[src/math/backends/webgl/lrn_gpu.ts, line 84 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2BzmApNf_0b9M-iFyK:-L2HMhpRPpqV-SOer5vz:b-6yb64q) ([raw file](https://github.com/pair-code/deeplearnjs/blob/02e1f519beb9e2bc908bd66c5603569f5a9baf74/src/math/backends/webgl/lrn_gpu.ts#L84)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Shouldn't the bounds of this be a function of arguments you splice in and not a function of anything at runtime?
</blockquote></details>

Unfortunately the loop condition depends on the current depth `d`, which is not static (and hence variable). Due to this, the shader fails when being parsed. If you have any idea how to get around this restriction I would appreciate any help to improve/fix this.

---

*[src/math/backends/webgl/lrn_gpu.ts, line 20 at r3](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2GudpROKCRoooXOqHD:-L2HZQk5vcNW97eHtjuB:b-xr9623) ([raw file](https://github.com/pair-code/deeplearnjs/blob/77965d0a67884c92310e6f661174c8f0430ac541/src/math/backends/webgl/lrn_gpu.ts#L20)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

this shader will get much simpler when we only support 4D input! :)
</blockquote></details>

I see what you mean ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2HaGwFS0nIYXod4mV5:b-s3zymp)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
try to do it like we do here for consistency:

https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/math.ts#L1670

 and then I will merge this PR. Nice work, the 4D shader alone is much easier to grok :)

---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/math/math.ts, line 2364 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHTiO6GB-eZ35rzcO:-L2IHTiO6GB-eZ35rzcP:b6vl4z7) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2364)):*
> ```XML
> 
>   /**
>    * Local response normalization - normalizes the activation of a local
> ```

You can remove "Local response normalization -" since it's the function name.

---

*[src/math/math.ts, line 2366 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHXT7aEVOE49rqxdC:-L2IHXT7aEVOE49rqxdD:b21fa1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2366)):*
> ```XML
>    * Local response normalization - normalizes the activation of a local
>    * neighborhood across or within channels.
>    * This is used for Caffe compatibility only. New models should use BatchNorm
> ```

Can you remove this? I think this may confuse people who are porting models (the second part)

---

*[src/math/math.ts, line 2392 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IG_MJPvliT8EqLWhl:-L2IG_MJPvliT8EqLWhm:b-86hez6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2392)):*
> ```XML
> 
>     return this.localResponseNormalization4D(
>                    x.reshape([1].concat(x.shape)) as Array4D, radius, bias,
> ```

Instead of calling reshape directly, do it like we do here: https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/math.ts#L1670

---

*[src/math/math.ts, line 2398 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHdENN3SHBJz1YUBW:-L2IHdENN3SHBJz1YUBX:b-2ztvz1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2398)):*
> ```XML
> 
>   /**
>    * Local response normalization - normalizes the activation of a local
> ```

same here, remove "local response normalization -"

---

*[src/math/math.ts, line 2400 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHfHFpJUpdut-QIjK:-L2IHfHFpJUpdut-QIjL:bh82c4y) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2400)):*
> ```XML
>    * Local response normalization - normalizes the activation of a local
>    * neighborhood across or within channels.
>    * This is used for Caffe compatibility only. New models should use BatchNorm
> ```

remove "new models should..."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-:-L2IGNwZu6Wc86qm6cF1:bem2u2z)*
<!-- Sent from Reviewable.io -->

chaosmail commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/math/math.ts, line 2364 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHTiO6GB-eZ35rzcO:-L2Jb7MwS7brj4mHKalP:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2364)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

You can remove "Local response normalization -" since it's the function name.
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2366 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHXT7aEVOE49rqxdC:-L2Jb8-Gv4NB0jYfA1tE:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2366)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Can you remove this? I think this may confuse people who are porting models (the second part)
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2392 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IG_MJPvliT8EqLWhl:-L2Jb95R7RFJ6zp1brt9:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2392)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Instead of calling reshape directly, do it like we do here: https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/math.ts#L1670
</blockquote></details>

Done.

---

*[src/math/math.ts, line 2400 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHfHFpJUpdut-QIjK:-L2Jb9pljPSxolBjc_2D:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2400)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove "new models should..."
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517)*
<!-- Sent from Reviewable.io -->

chaosmail commented:
Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/math/math.ts, line 2398 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517#-L2IHdENN3SHBJz1YUBW:-L2JcgCj2aVmxZxCLUT3:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3df08c82379891ec163fd42bc7792bec0b0786f2/src/math/math.ts#L2398)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

same here, remove "local response normalization -"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/517)*
<!-- Sent from Reviewable.io -->
